pageserver/src/tenant/manifest.rs

```diff
@@ -0,0 +1,252 @@
+    RemoveLayer(PersistentLayerDesc),
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct ManifestHeader {
+    version: usize,
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub enum Operation {
+    /// The header of the manifest (always the first record)
+    Header(ManifestHeader),
+    /// A snapshot of the current state
+    Snapshot(Snapshot, Lsn),
+    /// An atomic operation that changes the state
+    Operation(Vec<Record>, Lsn),
+}
+
+struct Header {
+    size: u32,
+    checksum: u32,
+}
+
+const HEADER_LEN: usize = 8;
+
+impl Header {
+    fn encode(&self) -> BytesMut {
+        let mut buf = BytesMut::with_capacity(HEADER_LEN);
+        buf.put_u32(self.size);
+        buf.put_u32(self.checksum);
+        buf
+    }
+
+    fn decode(mut buf: &[u8]) -> Self {
+        assert!(buf.len() == HEADER_LEN, "invalid header");
+        Self {
+            size: buf.get_u32(),
+            checksum: buf.get_u32(),
+        }
+    }
+}
+
+impl Manifest {
+    pub fn init(file: VirtualFile, snapshot: Snapshot, lsn: Lsn) -> Result<Self> {
+        let mut manifest = Self { file };
+        manifest.append_operation(Operation::Header(ManifestHeader { version: 1 }))?;
+        manifest.append_operation(Operation::Snapshot(snapshot, lsn))?;
+        Ok(manifest)
+    }
+
+    /// Load a manifest. Returns the manifest and a list of operations. If the manifest is corrupted,
+    /// the bool flag will be set to true and the user is responsible to reconstruct a new manifest and
+    /// backup the current one.
+    pub fn load(mut file: VirtualFile) -> Result<(Self, Vec<Operation>, bool)> {
+        let mut buf = vec![];
+        file.read_to_end(&mut buf)?;
+        let mut buf = Bytes::from(buf);
+        let mut operations = Vec::new();
+        let corrupted = loop {
+            if buf.remaining() == 0 {
+                break false;
+            }
+            if buf.remaining() < HEADER_LEN {
+                warn!("incomplete header when decoding manifest, could be corrupted");
+                break true;
+            }
+            let Header { size, checksum } = Header::decode(&buf[..HEADER_LEN]);
+            let size = size as usize;
+            buf.advance(HEADER_LEN);
+            if buf.remaining() < size {
+                warn!("incomplete data when decoding manifest, could be corrupted");
+                break true;
+            }
+            let data = &buf[..size];
+            if crc32c(data) != checksum {
+                warn!("checksum mismatch when decoding manifest, could be corrupted");
+                break true;
+            }
+            // if the following decode fails, we cannot use the manifest or safely ignore any record.
+            operations.push(serde_json::from_slice(data)?);
+            buf.advance(size);
+        };
+        let Operation::Header(header) = operations.remove(0) else {
+            bail!("cannot find manifest header");
+        };
```

Comment:
Should we also assert that any of the other operations are not headers? Or better yet, remove `Header` from the `Operation` enum so we don't have to.

Reply:
Yeah I plan to remove it from Operation enum